### PR TITLE
t/1509: Created an example of React used inside editor widgets

### LINKED ISSUE #1509: Create an example of React used inside the editor (in widgets)
It's possible to use any external library inside `UIElement`s so we can create a guide showcasing that.

### PR DESCRIPTION
**Pull request by oleq**
### Suggested merge commit message ([convention](https://github.com/ckeditor/ckeditor5-design/wiki/Git-commit-message-convention))

Docs: Created an example of React used inside editor widgets. Closes #1509.

### CODE REVIEW

**Comment by Reinmar** [comment]
😨 

This PR awaits my review so it's my lack of time why it's blocked that long. It should definitely be merged, though.

What could help is that if you confirmed that the guide that @oleq wrote is clear and works for you. This will give me the certainty that I don't have to go into details, so it will be a huge help.

**Comment by pomek** [comment]
I could help with the review but we need to resolve conflicts first. @oleq, could you?

**Comment by oleq** [comment]
Conflicts resolved. 

BTW, we should have a nice code highlight for editor content just like in the Link feature demo https://ckeditor5.github.io/docs/nightly/ckeditor5/latest/features/link.html#demo-2. I didn't know it's possible until @msamsel showed me this.

**Comment by msamsel** [comment]
I made it in a "hacky" way by using Umberto's features to highlight syntax in a way as Umberto is doing it.
However, this has some drawbacks. If Umberto starts to use a different highlighter or change way how it is done, then this "hack" will stop working.
Maybe it could be a nice feature to Umberto to expose highlighter in a proper way, maybe even with beautifier to make nice new lines and indentations.

**Comment by pomek** [comment]
I followed the guide. In order to work, I needed to change two things:

**style-loader** in `webpack.config.js`

```diff
{
	loader: 'style-loader',
	options: {
-		singleton: true
+		injectType: 'singletonStyleTag'
	}
},
```

**Import CSS** in `index.html`

```diff
<head>
    // ...
+   <link rel="stylesheet" href="assets/styles.css">
</head>
```

image

File: docs/framework/guides/tutorials/using-react-in-a-widget.md [hunk 1]
```diff
@@ -0,0 +1,1285 @@
...
+		// Feeding the application with pre–defined products.
+		// In a real-life application, this sort of data would be loaded
+		// from a database. To keep this tutorial simple, we're using
+		// a couple of hard–coded product definitions.
+		products={[
+			{
+				id: 1,
+				name: 'Colors of summer in Poland',
+				price: '$1500',
+				image: 'product1.jpg'
```
Comment (pomek) [inline docs/framework/guides/tutorials/using-react-in-a-widget.md]: Must the path be relative to the CSS file?

File: docs/framework/guides/tutorials/using-react-in-a-widget.md [hunk 2]
```diff
@@ -0,0 +1,1285 @@
...
+				query: {
+					presets: [ '@babel/react' ]
+				}
+			},
+			{
+				test: /ckeditor5-[^/\\]+[/\\]theme[/\\].+\.css/,
+				use: [
+					{
+						loader: 'style-loader',
+						options: {
```
[the remark this answers is not in the document]
Comment (oleq) [reply docs/framework/guides/tutorials/using-react-in-a-widget.md]: Yep. Style-loader API has changed. Options should be

```
options: {
	injectType: 'singletonStyleTag'
}
```

File: docs/framework/guides/tutorials/using-react-in-a-widget.md [hunk 3]
```diff
@@ -0,0 +1,1285 @@
...
+	babel-loader \
+	css-loader \
+	postcss-loader \
+	raw-loader \
+	react \
+	react-dom \
+	style-loader \
+	webpack \
+	webpack-cli
+```
```
[the remark this answers is not in the document]
Comment (oleq) [reply docs/framework/guides/tutorials/using-react-in-a-widget.md]: Yes. Check out the quick start guide and the `npm install` section https://github.com/ckeditor/ckeditor5/blob/master/docs/framework/guides/quick-start.md#lets-start.

File: docs/framework/guides/tutorials/using-react-in-a-widget.md (hunk 1 above)
Comment (oleq) [reply docs/framework/guides/tutorials/using-react-in-a-widget.md]: Not sure what's the problem here. Does not it work the way it is now?

File: docs/framework/guides/tutorials/using-react-in-a-widget.md [hunk 4]
```diff
@@ -0,0 +1,1285 @@
...
+bundle.js.map  1020 KiB    main  [emitted]         main
+Entrypoint main [big] = bundle.js bundle.js.map
+[./app.js] 391 bytes {main} [built]
+[./node_modules/webpack/buildin/global.js] (webpack)/buildin/global.js 472 bytes {main} [built]
+    + 11 hidden modules
+```
+
+You should see a "Hello world" application in your web browser, which might not be much but it is a good start:
+
+{@img assets/img/TODO.png TODO}
```
[the remark this answers is not in the document]
Comment (oleq) [reply docs/framework/guides/tutorials/using-react-in-a-widget.md]: You can use this one 

tutorial-react-in-widgets-hello-world

File: docs/framework/guides/tutorials/using-react-in-a-widget.md (hunk 1 above)
Comment (pomek) [reply docs/framework/guides/tutorials/using-react-in-a-widget.md]: Everything works fine. 

I just wanted to know why `assets/product1.jpg` which is a relative path to `index.html` didn't work. Then I figured out that the may the path must be relative to CSS file.

**Comment by pomek** [comment]
Ready for review once again.

**Comment by Reinmar** [comment]
> The guide is good -- just some grammatical nitpicks. Code is also good and example works as expected.

Thanks! Really appreciate the effort!



> I'm wondering if all these files (`webpack.config`, `index.html`, `app.js`, etc) should be in its own repo?

Oh, that would be great. But it's a problem already to maintain all the code samples (probably more than 100 of them already) and if we'd just place them in separate repos without some tooling for that it'd be a maintenance nightmare.  

In general, it's a PITA for people that code samples are hardly available. Sometimes they are just missing. Many times, we just copy/cut the code to the guide and clean it up so it's more readable and doesn't contain pieces specific for the documentation website (e.g. some config option). This is far from ideal, but seems to be the simpler option until we have some real solution for that in place.

I think that next year we'll look more into this. We also wanted to have the code available on some CodePen/Stackblitz/JSFiddle and for that we first need to clean up the situation with the need to rebuild the editor whenever you extend it.

**Comment by Reinmar** [comment]
Thanks, everyone! This amazing guide will finally be published 🎉

**Pull request merged**